jax/source_info_util.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextlib
+import inspect
+import os.path
+import threading
+from typing import List, NamedTuple, Optional
+
+class Frame(NamedTuple):
+  file_name: str
+  line_num: int
+  function_name: str
+
+SourceInfo = List[Frame]
+
+
+def _get_stacktrace() -> List[Frame]:
+  return [Frame(f.filename, f.lineno, f.function) for f in inspect.stack(0)]
+
+
+_jax_path = os.path.dirname(_get_stacktrace()[0].file_name)
```

Comment:
Isn't this the same as `os.dirname(__file__)`?

Reply:
I think so, yes!

Reply:
Though I'll defer to @hawkinsp to double check if he had a reason for writing it this way.

Reply:
Well, my concern was: what if the filename as reported by `__file__` has a different prefix to that I get from a stack trace? It's probably not a real problem; `__file__` is most likely fine.

Reply:
That seems like a good enough reason to keep it the way it is IMO.